src/Api/Helpers/AddDatabaseExtensionMethod.cs

```diff
@@ -47,6 +51,7 @@ public static IServiceCollection AddDatabase(this IServiceCollection services, I
 
             var environment = sp.GetRequiredService<IWebHostEnvironment>();
 
+            // need to add exception here to allow self hosted migrations or another way to initiate context

```

Comment:
comment 1: Currently, we cannot resolve the ITenantProvider when getting the DBTenantContext during SelfHosted Startup.

Not sure how to solve this in a clean way. Currently I feel like manually initiating it is the cleanest way, which is telling. @justindbaur Any ideas?

Or should we change the way we're applying migrations for self hosted production startup? We want to make it automatic.

Reply:
I have no idea what the best route forward is, so I laid out a few options that came to mind. Some easier, some that might be better architected. For sure not an exhaustive list though.

Option 1: Manually create the `DbTenantContext` by pushing in a fake `ITenantProvider`, the issue is this might not be as simple as you might like, I think you will be required to create the provider specific version. Otherwise migrations won't work, check on that though, I might be wrong. You'll have to do it like this basically. https://github.com/passwordless/passwordless-server/blob/1b9f4c17654739c5c7d00097e84456e19e021001/src/Service/Storage/EF/DbTenantFactory.cs#L22

Option 2: Install the `dotnet-ef` tool in the docker file and run `dotnet ef database update` based on user configuration in `entrypoint.sh`. I'm not actually sure if this is possible, because you'd be doing this on a `Passwordless.Api.dll` and not on the csproj. That command has a `--no-build` command but I think this is still stretching possibilities. Worth investigating further though.

Option 3: Do it in a IHostedService with it's own service provider scope, then run migrations. (This could be done in startup too)

Option 4: Totally move away from the `DbTenantContext` being responsible for the tenant separation and instead it's "optional" and enforced on a per query basis. 

Option 5: Have 2 DbContext's, one that is more management, and does all the stuff that is behind the ManagementKey checks, and one that is tenant scoped. The management one is the one you use to do migrations and such. The downside is that you have to implement some DbSets multiple times.

Reply:
Thanks for all the input @justindbaur, I will experiment a bit and see if I can cook something up that looks OK and then we'll go from there.